File: crates/ecr-store/src/markdown.rs
//! Reading a message's HTML as text.
//!
//! The plain-text view used to be whatever `mail_parser::body_text` answered.
//! For a message with no `text/plain` part that is the parser's own flattening
//! of the markup, which runs block elements together — `<div>one</div>
//! <div>two</div>` arrives as `onetwo` — and for a message that has one it is
//! usually the `text/plain` half of a marketing alternative, which is a line
//! saying the mail cannot be displayed and a URL. Neither is a reading of the
//! message.
//!
//! So the text view is the markup, converted: headings, emphasis, lists, quotes
//! and links survive as the punctuation they were always written as, and the
//! result is legible in a monospaced pane with nothing rendering it. The real
//! `text/plain` part is the fallback rather than the first choice, on the
//! grounds that a sender who wrote HTML wrote the HTML — and it is still what
//! answers when there is no markup at all.
//!
//! Nothing here is a security boundary. The output is inserted as text, so
//! what `<script>` becomes does not matter; it is dropped because a reader
//! does not want to read it.
//!
//! The one exception is an image, which the client turns back into an `<img>`
//! — a `src` is an attribute rather than text, and it is the only thing here
//! that reaches a DOM as anything but a string. What it is allowed to point at
//! is decided twice and in both languages: `mime::rewrite_markdown_images`
//! resolves `cid:` and applies the reader's remote-images setting, and
//! `ui/linkify.ts` refuses any scheme that is not a way of fetching a picture.

use std::sync::OnceLock;

/// Tags whose *contents* are not the message.
///
/// `img` used to be on this list, and the reason it no longer is is worth
/// writing down, because the original reason was sound. htmd renders an image
/// as `![alt](src)`, and real mail is tracking pixels, spacer gifs and
/// sliced-up letterheads with no alt text — so a reader who is shown the
/// *markdown* gets a column of `![](https://…)` between the sentences, which
/// is why they were dropped.
///
/// What changed is that the client no longer shows the markdown. It renders
/// it: `![alt](src)` becomes an `<img>`, so a spacer gif is a spacer gif and a
/// letterhead is a letterhead, exactly as in the HTML view. The complaint was
/// never about images, it was about URLs standing in for them.
///
/// `svg` stays out. It is markup rather than a resource, it does not survive
/// the round trip through markdown as anything a client can render, and it is
/// the one image format that can carry script.
const NOT_THE_MESSAGE: &[&str] = &[
    "script", "style", "head", "title", "meta", "link", "noscript", "iframe", "svg",
];

fn converter() -> &'static htmd::HtmlToMarkdown {
    static CONVERTER: OnceLock<htmd::HtmlToMarkdown> = OnceLock::new();
    CONVERTER.get_or_init(|| {
        htmd::HtmlToMarkdown::builder()
            .skip_tags(NOT_THE_MESSAGE.to_vec())
            .options(htmd::options::Options {
                // A dash and one space, rather than the default asterisk and
                // three. This is read in a proportional pane beside prose, not
                // in an editor: `- item` is a list, `*   item` is a column of
                // punctuation with a gap after it. Emphasis is still asterisks,
                // so a leading `*` would be the one marker that means two
                // things.
                bullet_list_marker: htmd::options::BulletListMarker::Dash,
                ul_bullet_spacing: 1,
                ol_number_spacing: 1,
                ..Default::default()
            })
            .build()
    })
}

/// `None` when the conversion failed or came back with nothing in it, so the
/// caller can fall back rather than show an empty pane.
pub fn from_html(html: &str) -> Option<String> {
    let markdown = converter().convert(html).ok()?;
    let trimmed = markdown.trim();

    (!trimmed.is_empty()).then(|| trimmed.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn block_elements_do_not_run_together() {
        let out = from_html("<div>one</div><div>two</div>").expect("markdown");
        assert_eq!(out, "one\n\ntwo");
    }

    #[test]
    fn structure_survives_as_punctuation() {
        let out = from_html(
            "<h1>Notice</h1><p>Please <strong>read</strong> this.</p>\
             <ul><li>first</li><li>second</li></ul>",
        )
        .expect("markdown");

        assert!(out.contains("# Notice"), "{out}");
        assert!(out.contains("**read**"), "{out}");
        assert!(out.contains("- first"), "{out}");
    }

    #[test]
    fn a_link_keeps_both_its_words_and_its_url() {
        let out = from_html(r#"<p>See <a href="https://example.com/x">the notes</a>.</p>"#)
            .expect("markdown");

        assert_eq!(out, "See [the notes](https://example.com/x).");
    }

    /// A stylesheet is not a message, and mail carries kilobytes of it.
    #[test]
    fn styles_and_scripts_are_not_read_out() {
        let out = from_html("<style>.a{color:red}</style><script>alert(1)</script><p>Hello.</p>")
            .expect("markdown");

        assert_eq!(out, "Hello.");
    }

    /// The client renders these back into `<img>`, so the src has to survive
    /// the conversion rather than being dropped as noise.
    #[test]
    fn an_image_keeps_its_source() {
        let out = from_html(
            r#"<p>Hello.</p><img src="https://t.example.com/pixel.gif" width="1" height="1">"#,
        )
        .expect("markdown");

        assert!(
            out.contains("![](https://t.example.com/pixel.gif)"),
            "{out}"
        );
    }

    /// The one image format that can carry script, and the one that does not
    /// survive the round trip as anything renderable.
    #[test]
    fn inline_svg_is_still_not_the_message() {
        let out =
            from_html(r#"<p>Hello.</p><svg><script>alert(1)</script></svg>"#).expect("markdown");

        assert_eq!(out, "Hello.");
    }

    #[test]
    fn markup_with_nothing_in_it_answers_none() {
        assert_eq!(from_html("<style>.a{color:red}</style>"), None);
        assert_eq!(from_html(""), None);
    }
}

File: web/src/ui/markdown.ts
/**
 * The reading text, rendered.
 *
 * `BodyFormat::Text` is the message's markup read as Markdown — see
 * `ecr_store::markdown` — and it used to be shown exactly as it arrived, on the
 * grounds that Markdown is legible without anything rendering it. It is, up to
 * a point, and the point is emphasis: `Rich **HTML** body` is not a sentence
 * anyone wants to read, and `[the notes](https://…/a/very/long/path)` is a URL
 * wedged into the middle of one. The punctuation stands in for something rather
 * than decorating it, which is the same reason images had to be rendered.
 *
 * So this renders the marks and keeps everything else. The pane stays a
 * monospaced `<pre>` with the source's own newlines: nothing here emits a block
 * element, so a line is still a line, `j`/`k` in view mode still move by one,
 * and the flat text view is still recognisably the flat text view rather than a
 * second copy of the HTML one.
 *
 * What the converter actually emits is the whole of what is handled, and it is
 * a short list: `**strong**`, `*em*`, `` `code` ``, `#` headings, `>` quotes,
 * `- `/`1. ` list markers, `* * *` rules and ``` fences. There is no
 * strikethrough because htmd drops `<s>` entirely, and no tables because it
 * writes each cell as its own paragraph.
 *
 * Only two things here reach the DOM as anything but text — an image's `src`
 * and a link's `href` — and both are checked against what they are allowed to
 * be. Everything else is escaped.
 */
import { escapeHtml, linkify } from "./linkify";

/**
 * What an image is allowed to point at.
 *
 * The server has already resolved `cid:` to a part of this message and dropped,
 * or kept, whatever was remote according to the reader's setting — so a target
 * arriving here was meant to be fetched. This is the second layer.
 *
 * `data:image/svg+xml` is refused with the rest. An `<img>` does not run script
 * in an SVG it loads, so that is safe by specification rather than by
 * construction, and nothing in mail needs it.
 */
const FETCHABLE = /^(?:https?:\/\/|\/(?!\/)|data:image\/(?!svg))/i;

/**
 * What a link is allowed to point at.
 *
 * Wider than `FETCHABLE`, because `mailto:` is a thing a message legitimately
 * says and `ui/follow-link.ts` turns one into a composer. Narrower than
 * everything, because `javascript:` is a scheme.
 */
const FOLLOWABLE = /^(?:https?:\/\/|mailto:|\/(?!\/))/i;

/** htmd writes a `<pre><code>` block between these. */
const FENCE = /^\s*```/;
const HEADING = /^(#{1,6})\s+(.*)$/;
/** And an `<hr>` as this. Tested before the bullet, which it would also match. */
const RULE = /^\s*(?:\*\s?){3,}\s*$/;
const QUOTE = /^((?:>\s?)+)(.*)$/;
/**
 * A list marker.
 *
 * The space after it is what keeps `*emphasis*` out: a marker is followed by
 * the item, an opening delimiter is followed by the word it opens.
 */
const BULLET = /^(\s*)([-*+]|\d+[.)])\s+(.*)$/;

/**
 * Everything inline, in one ordered alternation.
 *
 * One pass rather than one per construct, because scanning separately means
 * whichever ran second sees the other's output — `![alt](url)` found again as a
 * bare `!` beside a link, `**bold**` found again as two empty emphases. The
 * order is the precedence: a backtick span is literal, so it wins over
 * everything inside it, and `**` is tried before `*` at the same position.
 *
 * `(?=\S)` on both emphases is what keeps arithmetic and prose out of it: an
 * opening delimiter is followed by the word it opens, never by a space, so
 * `2 * 3 * 4` is multiplication and stays that way.
 */
const INLINE = new RegExp(
	[
		"`([^`\\n]+)`", // 1 code
		"(!?)\\[([^\\]\\n]*)\\]\\(([^)\\s]+)\\)", // 2 bang, 3 label, 4 target
		"\\*\\*(?=\\S)([^\\n]*?\\S)\\*\\*", // 5 strong
		"\\*(?=\\S)([^*\\n]*?\\S|\\S)\\*", // 6 em
	].join("|"),
	"g",
);

/**
 * How far emphasis may nest before it is read as text.
 *
 * Message bodies are untrusted, and the inner content of a match is rendered by
 * the same function that found it. Real mail nests two deep at the most; a run
 * of asterisks contrived to nest a thousand is not a message.
 */
const MAX_DEPTH = 6;

export function renderMarkdown(text: string): string {
	const out: string[] = [];
	let fenced: string[] | null = null;

	for (const line of text.split("\n")) {
		if (FENCE.test(line)) {
			if (fenced) {
				out.push(fence(fenced));
				fenced = null;
			} else {
				fenced = [];
			}
			continue;
		}

		if (fenced) {
			fenced.push(line);
			continue;
		}

		out.push(renderLine(line));
	}

	// A fence that never closes is a message that was cut short, not a reason to
	// drop everything after the last backticks.
	if (fenced) out.push(fence(fenced));

	return out.join("\n");
}

const fence = (lines: string[]) =>
	`<code class="md-fence">${escapeHtml(lines.join("\n"))}</code>`;

function renderLine(line: string): string {
	if (RULE.test(line)) return '<span class="md-rule" aria-hidden="true"></span>';

	const heading = HEADING.exec(line);
	if (heading) {
		const level = heading[1]!.length;
		return `<span class="md-h md-h${level}">${renderInline(heading[2] ?? "")}</span>`;
	}

	const quote = QUOTE.exec(line);
	if (quote) {
		// The marker is kept. It is the only thing left saying where the quote
		// ends, in a pane that cannot draw a rule down the side of one.
		return `<span class="md-quote">${escapeHtml(quote[1] ?? "")}${renderInline(quote[2] ?? "")}</span>`;
	}

	const bullet = BULLET.exec(line);
	if (bullet) {
		return `${bullet[1]}<span class="md-marker">${escapeHtml(bullet[2]!)}</span> ${renderInline(bullet[3] ?? "")}`;
	}

	return renderInline(line);
}

/**
 * `autolink` is off inside a link's own words: a bare URL there would be turned
 * into a second anchor inside the first, which is not markup a browser can
 * make sense of. It is exactly the case `[](https://…)` produces, since an
 * empty label falls back to showing the target.
 */
function renderInline(text: string, depth = 0, autolink = true): string {
	const plain = (value: string) => (autolink ? linkify(value) : escapeHtml(value));
	if (depth >= MAX_DEPTH) return plain(text);

	let out = "";
	let at = 0;

	for (const match of text.matchAll(INLINE)) {
		const index = match.index ?? 0;
		out += plain(text.slice(at, index));

		if (match[1] !== undefined) {
			out += `<code>${escapeHtml(match[1])}</code>`;
		} else if (match[4] !== undefined) {
			out += target(match[2] === "!", match[3] ?? "", match[4], match[0], depth);
		} else if (match[5] !== undefined) {
			out += `<strong>${renderInline(match[5], depth + 1, autolink)}</strong>`;
		} else if (match[6] !== undefined) {
			out += `<em>${renderInline(match[6], depth + 1, autolink)}</em>`;
		}

		at = index + match[0].length;
	}

	return out + plain(text.slice(at));
}

/**
 * An image or a link, or neither.
 *
 * A refused *image* falls back to its alt text, which is all it ever said. A
 * refused *link* is left as the text it was written as: a reader being shown
 * something odd is better off seeing where it claimed to point, and it cannot
 * be clicked.
 */
function target(
	image: boolean,
	label: string,
	href: string,
	raw: string,
	depth: number,
): string {
	if (image) {
		return FETCHABLE.test(href)
			? `<img class="body-image" src="${escapeHtml(href)}" alt="${escapeHtml(label)}" loading="lazy">`
			: escapeHtml(label);
	}

	if (!FOLLOWABLE.test(href)) return escapeHtml(raw);

	const words = label ? renderInline(label, depth + 1, false) : escapeHtml(href);
	return `<a href="${escapeHtml(href)}" target="_blank" rel="noreferrer noopener">${words}</a>`;
}
